hash_practice/exercises.py

```diff
@@ -22,8 +40,22 @@ def valid_sudoku(table):
         Each element can either be a ".", or a digit 1-9
         The same digit cannot appear twice or more in the same 
         row, column or 3x3 subgrid
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n2)
+        Space Complexity: O(1)
     """
-    pass
+    for row in table: 
+        for element in row: 
+            if element != ".":
+                if element not in "123456789":
+                    return False 
+                else: 
+                    if row.count(element) >1: 
+                        return False 
+                    if [row[i] for i in range(9)].count(element)>1:
+                        return False 
+                    if [table[i][j] for i in range(3) for j in range(3)].count(element) >1: 

```

Comment:
Likewise, this will always return the first subgrid ((0, 0) to (2, 2)) every iteration, so to fully solve this, you will likely need to more explicitly iterate through all the subgrids rather than using a list comprehension.